subprojects/model-core/src/main/java/org/gradle/api/internal/provider/DefaultProperty.java

```diff
@@ -59,7 +59,11 @@ public void setFromAnyValue(Object object) {
         if (object instanceof Provider) {
             set(Cast.<Provider<T>>uncheckedNonnullCast(object));
         } else {
-            set(Cast.<T>uncheckedNonnullCast(object));
+            if (object != null && type.isAssignableFrom(String.class) && !(object instanceof String)) {
+                set(Cast.<T>uncheckedNonnullCast(String.valueOf(object)));
```

Comment:
Why not call `toString()` directly?

What happens with exceptions here? It would be good to add a test to see how a problem with a `toString()` method is reported here.

Reply:
Yeah, calling `toString()` directly makes sense (after a null check). I agree this change needs more tests if we decide to merge it.